wpilibj/src/test/java/edu/wpi/first/wpilibj/event/BooleanEventTest.java

```diff
@@ -58,15 +59,89 @@ void testBinaryCompositionLoopSemantics() {
 
     assertEquals(1, counter1.get());
     assertEquals(1, counter2.get());
+
+    bool2.set(false);
+    loop1.poll();
+
+    assertEquals(2, counter1.get());
+    assertEquals(1, counter2.get());
+
+    loop2.poll();
+
+    assertEquals(2, counter1.get());
+    assertEquals(1, counter2.get());
+
+    loop1.poll();
+
+    assertEquals(2, counter1.get());
+    assertEquals(1, counter2.get());
+
+    bool2.set(true);
+    loop2.poll();
+
+    assertEquals(2, counter1.get());
+    assertEquals(1, counter2.get());
```

Comment:
Why shouldn't `counter2` be incremented?

It's quite difficult to understand what's being tested, how, and why (from a code flow/logic standpoint, not a design one) whatever is being asserted is indeed what should happen -- other cases took me some thinking, and this one baffles me. 

Inline comments explaining the test flow and why each assertion is the correct behavior would be very helpful.